plugin/forward/dnstap.go

```diff
@@ -34,33 +34,37 @@ func toDnstap(ctx context.Context, f *Forward, host string, state request.Reques
 	}
 
 	for _, t := range f.tapPlugins {
-		// Query
-		q := new(tap.Message)
-		msg.SetQueryTime(q, start)
-		// Forwarder dnstap messages are from the perspective of the downstream server
-		// (upstream is the forward server)
-		msg.SetQueryAddress(q, state.W.RemoteAddr())
-		msg.SetResponseAddress(q, ta)
-		if t.IncludeRawMessage {
-			buf, _ := state.Req.Pack()
-			q.QueryMessage = buf
+		if t.MessageTypeEnabled(tap.Message_FORWARDER_QUERY) {
```

Comment:
It might make sense to check the enabled message types at very begin of this function and save resources on processing hosts, ports, etc. Or even better, you can check the types in
https://github.com/coredns/coredns/blob/a7ed346585e30b99317d36e4d007b7b19a228ea5/plugin/forward/forward.go#L76-L81
i.e. skip the tap plugin, it both FQ and FR messages are disabled

Reply:
https://github.com/coredns/coredns/pull/6643/commits/109b2048ad82c8ecf5715e32f5f1347e8eb22ed4